src/Parsers/ExpressionListParsers.cpp

```diff
@@ -1042,4 +1107,1887 @@ bool ParserKeyValuePairsList::parseImpl(Pos & pos, ASTPtr & node, Expected & exp
+            auto cur_op = op_table_unary.begin();
+            for (; cur_op != op_table_unary.end(); ++cur_op)
+            {
+                if (parseOperator(pos, cur_op->first, expected))
+                    break;
+            }
+
+            if (cur_op != op_table_unary.end())
+            {
+                next = Action::OPERAND;
+                layers.back()->pushOperator(cur_op->second);
+                continue;
+            }
+
+            auto old_pos = pos;
+            std::unique_ptr<Layer> layer;
+            if (parseOperator(pos, "INTERVAL", expected))
+                layer = std::make_unique<IntervalLayer>();
+            else if (parseOperator(pos, "CASE", expected))
+                layer = std::make_unique<CaseLayer>();
+
+            /// Here we check that CASE or INTERVAL is not an identifier
+            /// It is needed for backwards compatibility
+            if (layer)
+            {
+                Expected stub;
+
+                auto stub_cur_op = op_table.begin();
+                for (; stub_cur_op != op_table.end(); ++stub_cur_op)
+                {
+                    /// Minus can be unary
+                    /// TODO: check cases 'select case - number from table' and 'select case -x when 10 then 5 else 0 end'
+                    if (stub_cur_op->second.function_name == "minus")
+                        continue;
+                    if (parseOperator(pos, stub_cur_op->first, stub))
+                        break;
+                }
+
+                auto check_pos = pos;
+
+                if (stub_cur_op != op_table.end() ||
+                    ParserToken(TokenType::Comma).ignore(pos, stub) ||
+                    ParserToken(TokenType::ClosingRoundBracket).ignore(pos, stub) ||
+                    ParserToken(TokenType::ClosingSquareBracket).ignore(pos, stub) ||
+                    ParserToken(TokenType::Semicolon).ignore(pos, stub) ||
+                    ParserKeyword("AS").ignore(pos, stub) ||
+                    ParserKeyword("FROM").ignore(pos, stub) ||
+                    !pos.isValid())
+                {
+                    pos = old_pos;
+                }
+                else if (ParserAlias(true).ignore(check_pos, stub) &&
+                         (ParserToken(TokenType::Comma).ignore(check_pos, stub) ||
+                          ParserToken(TokenType::ClosingRoundBracket).ignore(check_pos, stub) ||
+                          ParserToken(TokenType::ClosingSquareBracket).ignore(check_pos, stub) ||
+                          ParserToken(TokenType::Semicolon).ignore(check_pos, stub) ||
+                          ParserKeyword("FROM").ignore(check_pos, stub) ||
+                          !check_pos.isValid()))
+                {
+                    pos = old_pos;
+                }
+                else
+                {
+                    next = Action::OPERAND;
+                    layers.push_back(std::move(layer));
+                    continue;
+                }
```

Comment:
It looks scary.
Maybe we can support some kind of rollback here?
Parse INTERVAL or CASE as operators first. If we have not managed to parse current layer, ten rollback ad try it as identifier.

Reply:
Just, it is not clear if we have covered all the cases here.
Maybe it's ok for this specific scenario, but looks suspicious.